### PR TITLE
Repository maintenance support

### PR DESCRIPTION
**Pull request by jkowalski**
Moved maintenance algorithms from cli/ to repo/maintenance/

Added support for CLI commands:

`kopia gc` - performs quick maintenance
`kopia gc --full`- perform full maintenance

Full maintenance performs `snapshot gc`, but it's not safe to do this automatically possibly in parallel to snapshots being taken. This will be addressed ~0.7 timeframe.

### CODE REVIEW

File: cli/command_maintenance_info.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/pkg/errors"
```
Comment (julio-lopez) [inline cli/command_maintenance_info.go:7]:
```suggestion
	"github.com/pkg/errors"

	"github.com/kopia/kopia/repo"
```

File: cli/command_maintenance_set.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package cli
+
+import (
+	"context"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/pkg/errors"
```
Comment (julio-lopez) [inline cli/command_maintenance_set.go:7]:
```suggestion
	"github.com/pkg/errors"

	"github.com/kopia/kopia/repo"
```

File: cli/command_maintenance_set.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+var (
+	maintenanceSetCommand     = maintenanceCommands.Command("set", "Set maintenance parameters")
+	maintenanceSetFrequency   = maintenanceSetCommand.Flag("frequency", "Set garbage collection frequency").DurationList()
+	maintenanceSetEnabled     = maintenanceSetCommand.Flag("enabled", "Enable garbage collection").BoolList()
+	maintenanceSetNextRunTime = maintenanceSetCommand.Flag("next-run-time", "Śet next run time").String()
+)
+
+func runGCSetParams(ctx context.Context, rep *repo.DirectRepository) error {
```
Comment (julio-lopez) [inline cli/command_maintenance_set.go:17]:
```suggestion
func runGCSetParams(ctx context.Context, rep *repo.DirectRepository) error {
	// silence linter for now
	_ = maintenanceSetEnabled
	_ = maintenanceSetFrequency
	_ = maintenanceSetNextRunTime

```

File: cli/command_maintenance_set.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+	"context"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/pkg/errors"
+)
+
+var (
+	maintenanceSetCommand     = maintenanceCommands.Command("set", "Set maintenance parameters")
+	maintenanceSetFrequency   = maintenanceSetCommand.Flag("frequency", "Set garbage collection frequency").DurationList()
+	maintenanceSetEnabled     = maintenanceSetCommand.Flag("enabled", "Enable garbage collection").BoolList()
```
Comment (julio-lopez) [inline cli/command_maintenance_set.go:13]:
```suggestion
	maintenanceSetEnabled     = maintenanceSetCommand.Flag("enabled", "Enable garbage collection").BoolList()
	maintenanceSetFrequency   = maintenanceSetCommand.Flag("frequency", "Set garbage collection frequency").DurationList()
```

File: repo/maintenance/maintenance_manager.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	// as deleted for a long time
+	if err := m.DropDeletedContents(ctx, -1); err != nil {
+		return errors.Wrap(err, "error dropping deleted contents")
+	}
+
+	// find 'q' packs that are less than 80% full and rewrite contents in them into
+	// new consolidated packs, orphaning old packs in the process.
+	if err := m.RewriteContents(ctx, &RewriteContentsOptions{
+		ContentIDRange: content.AllPrefixedIDs,
+		PackPrefix:     content.PackBlobIDPrefixRegular,
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply repo/maintenance/maintenance_manager.go]: good catch.

**Comment by jkowalski** [comment]
I re-did the entire PR and squashed to one commit. Instead of maintenance.Manager, it's now just a functional package.

`maintenance.Params` is stored in a manifest with `type:maintenance`
`maintenance.Schedule` is stored in the root of a repository, unencrypted but that should be fine since it has no sensitive data only time when the next maintenance should run.

**Comment by julio-lopez** [comment]
OK, this means reviewing the entire PR again.

File: tests/end_to_end_test/snapshot_delete_test.go [hunk 6]
```diff
@@ -252,21 +252,26 @@ func TestSnapshotDeleteTypeCheck(t *testing.T) {
 	e.RunAndExpectSuccess(t, "repo", "create", "filesystem", "--path", e.RepoDir)
 
 	lines := e.RunAndExpectSuccess(t, "manifest", "ls")
-	if len(lines) != 1 {
-		t.Fatalf("Expected 1 line global policy output for manifest ls")
+	if len(lines) != 2 {
```
Comment (julio-lopez) [inline tests/end_to_end_test/snapshot_delete_test.go:255]: @redgoat650 FYI

**Pull request merged**